Incorrect generic parameter ordering for E0747

https://github.com/rust-lang/rust/blob/a59264b01247836c70e24217e0d346b868387525/src/test/ui/suggestions/suggest-move-types.stderr#L121-L135
suggests type parameters should come before lifetime parameters, which is incorrect.
The message is emitted here:
https://github.com/rust-lang/rust/blob/a59264b01247836c70e24217e0d346b868387525/src/librustc_typeck/astconv.rs#L463-L477

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->
This issue has been assigned to @camelid via [this comment](https://github.com/rust-lang/rust/issues/72815#issuecomment-636506152).
<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"camelid"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

I've been trying to fix this, but even reversing the arguments to the `err.note` format didn't seem to do anything. Am I missing something?

I also added `ABC` to the error note for testing:

```rust
err.note(&format!("{} arguments must be provided before {} arguments ABC", arg.descr(), kind));
```

And here's the error diff:

```
---- [ui] ui/suggestions/suggest-move-types.rs stdout ----
diff of stderr:

108	LL | struct Al<'a, T, M: OneWithLifetime<A=(), T, 'a>> {
109	   |                                           ^
110	   |
-	   = note: lifetime arguments must be provided before type arguments
+	   = note: type arguments must be provided before lifetime arguments ABC
112	
113	error[E0747]: type provided when a lifetime was expected
114	  --> $DIR/suggest-move-types.rs:48:71

116	LL | struct Bl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<A=(), B=(), C=(), T, U, V, 'a, 'b, 'c>> {
117	   |                                                                       ^
118	   |
-	   = note: lifetime arguments must be provided before type arguments
+	   = note: type arguments must be provided before lifetime arguments ABC
120	
121	error[E0747]: lifetime provided when a type was expected
122	  --> $DIR/suggest-move-types.rs:65:56

124	LL | struct Cl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), C=(), U, 'b, V, 'c>> {
125	   |                                                        ^^
126	   |
-	   = note: lifetime arguments must be provided before type arguments
+	   = note: lifetime arguments must be provided before type arguments ABC
128	
129	error[E0747]: lifetime provided when a type was expected
130	  --> $DIR/suggest-move-types.rs:82:56

132	LL | struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), U, 'b, C=(), V, 'c>> {
133	   |                                                        ^^
134	   |
-	   = note: lifetime arguments must be provided before type arguments
+	   = note: lifetime arguments must be provided before type arguments ABC
136	
137	error: aborting due to 12 previous errors
138	


The actual stderr differed from the expected stderr.
```

--- a/src/librustc_typeck/astconv.rs
+++ b/src/librustc_typeck/astconv.rs
@@ -491,7 +491,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             "constant" => ParamKindOrd::Const,
             // It's more concise to match on the string representation, though it means
             // the match is non-exhaustive.
-            _ => bug!("invalid generic parameter kind"),
+            _ => bug!("invalid generic parameter kind {}", kind),
         };
         let arg_ord = match arg {
             GenericArg::Lifetime(_) => ParamKindOrd::Lifetime,

--- a/src/librustc_typeck/astconv.rs
+++ b/src/librustc_typeck/astconv.rs
@@ -489,7 +489,9 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             "lifetime" => ParamKindOrd::Lifetime,
             "type" => ParamKindOrd::Type,
             "constant" => ParamKindOrd::Const,
-            _ => panic!(),
+            // It's more concise to match on the string representation, though it means
+            // the match is non-exhaustive.
+            _ => bug!("invalid generic parameter kind"),
         };
         let arg_ord = match arg {
             GenericArg::Lifetime(_) => ParamKindOrd::Lifetime,

--- a/src/librustc_typeck/astconv.rs
+++ b/src/librustc_typeck/astconv.rs
@@ -8,6 +8,7 @@
 use crate::collect::PlaceholderHirTyCollector;
 use crate::middle::resolve_lifetime as rl;
 use crate::require_c_abi_if_c_variadic;
+use rustc_ast::ast::ParamKindOrd;
 use rustc_ast::util::lev_distance::find_best_match_for_name;
 use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_errors::ErrorReported;
@@ -483,8 +484,23 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             arg.descr(),
             kind,
         );
+
+        let kind_ord = match kind {
+            "lifetime" => ParamKindOrd::Lifetime,
+            "type" => ParamKindOrd::Type,
+            "constant" => ParamKindOrd::Const,
+            _ => panic!(),
+        };
+        let arg_ord = match arg {
+            GenericArg::Lifetime(_) => ParamKindOrd::Lifetime,
+            GenericArg::Type(_) => ParamKindOrd::Type,
+            GenericArg::Const(_) => ParamKindOrd::Const,
+        };
+
         // This note will be true as long as generic parameters are strictly ordered by their kind.
-        err.note(&format!("{} arguments must be provided before {} arguments", kind, arg.descr()));
+        let (first, last) =
+            if kind_ord < arg_ord { (kind, arg.descr()) } else { (arg.descr(), kind) };
+        err.note(&format!("{} arguments must be provided before {} arguments", first, last));
         err.emit();
     }
 

--- a/src/test/ui/suggestions/suggest-move-types.stderr
+++ b/src/test/ui/suggestions/suggest-move-types.stderr
@@ -124,7 +124,7 @@ error[E0747]: lifetime provided when a type was expected
 LL | struct Cl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), C=(), U, 'b, V, 'c>> {
    |                                                        ^^
    |
-   = note: type arguments must be provided before lifetime arguments
+   = note: lifetime arguments must be provided before type arguments
 
 error[E0747]: lifetime provided when a type was expected
   --> $DIR/suggest-move-types.rs:82:56
@@ -132,7 +132,7 @@ error[E0747]: lifetime provided when a type was expected
 LL | struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), U, 'b, C=(), V, 'c>> {
    |                                                        ^^
    |
-   = note: type arguments must be provided before lifetime arguments
+   = note: lifetime arguments must be provided before type arguments
 
 error: aborting due to 12 previous errors
 

